File: conformance/reference-js/response-mapping.js
'use strict';

function failure(code, path) {
  return { ok: false, code, path };
}

function readPath(response, path) {
  let current = response;
  for (const segment of path.split('.')) {
    if (current === null || typeof current !== 'object' || Array.isArray(current)
      || !Object.prototype.hasOwnProperty.call(current, segment)) {
      return { found: false };
    }
    current = current[segment];
  }
  return { found: true, value: current };
}

function resolveMappedValue(response, path) {
  const resolved = readPath(response, path);
  return resolved.found ? resolved : failure('RESPONSE_MAPPING_PATH_MISSING', path);
}

function wireTypeMatches(wire, value) {
  if (wire === 'string') return typeof value === 'string';
  if (wire === 'boolean') return typeof value === 'boolean';
  if (wire === 'number') return typeof value === 'number' && Number.isFinite(value);
  if (wire === 'array') return Array.isArray(value);
  return true;
}

function mapFormRecord(input) {
  const mapping = input.responseMapping;
  if (mapping === undefined || mapping === null || typeof mapping !== 'object'
    || Array.isArray(mapping) || Object.keys(mapping).length === 0) {
    return failure('INVALID_RESPONSE_MAPPING', 'responseMapping');
  }
  const values = {};
  const skipped = {};
  const fieldWireTypes = input.fieldWireTypes && typeof input.fieldWireTypes === 'object'
    && !Array.isArray(input.fieldWireTypes)
    ? input.fieldWireTypes
    : null;
  for (const [field, pathExpr] of Object.entries(mapping)) {
    if (typeof field !== 'string' || field.length === 0
      || typeof pathExpr !== 'string' || pathExpr.length === 0) {
      return failure('INVALID_RESPONSE_MAPPING', `responseMapping.${field}`);
    }
    const resolved = readPath(input.response, pathExpr);
    // ADR-0021 D5a / V273: missing path does not abort the whole fill.
    // Conformance-observable value is JSON null (equivalent empty initial; not undefined).
    if (!resolved.found) {
      values[field] = null;
      continue;
    }
    const wire = fieldWireTypes ? fieldWireTypes[field] : undefined;
    if (wire && !wireTypeMatches(wire, resolved.value)) {
      // ADR-0028 OQ-1: type mismatch skips the key; does not abort whole fill.
      skipped[field] = 'FIELD_WIRE_TYPE_MISMATCH';
      continue;
    }
    values[field] = resolved.value;
  }
  const result = { ok: true, values };
  if (Object.keys(skipped).length > 0) result.skipped = skipped;
  return result;
}

function mapResponse(input) {
  // formRecord (ADR-0021) and recordView (ADR-0024) share explicit mapping → values.
  if (input.component === 'formRecord' || input.component === 'recordView') {
    return mapFormRecord(input);
  }
  const mapping = Object.hasOwn(input, 'localMapping')
    ? input.localMapping
    : input.datasourceMapping;
  if (mapping !== undefined && (mapping === null || typeof mapping !== 'object' || Array.isArray(mapping) || Object.keys(mapping).length === 0)) {
    return failure('INVALID_RESPONSE_MAPPING', 'localMapping');
  }
  if ((input.component === 'statCard' || input.component === 'text') && mapping !== undefined) {
    return failure('RESPONSE_MAPPING_NOT_SUPPORTED', 'localMapping');
  }
  if (input.component === 'chart' && mapping === undefined) {
    return Array.isArray(input.response)
      ? { ok: true, data: { list: input.response } }
      : failure('RESPONSE_MAPPING_TYPE_MISMATCH', '$');
  }

  const listPath = mapping?.list || 'list';
  const list = resolveMappedValue(input.response, listPath);
  if (!list.found) return list;
  if (!Array.isArray(list.value)) return failure('RESPONSE_MAPPING_TYPE_MISMATCH', listPath);

  const data = { list: list.value };
  if (input.component === 'table' && input.paginationMode === 'server') {
    const totalPath = mapping?.total || 'total';
    const total = resolveMappedValue(input.response, totalPath);
    if (!total.found) return total;
    if (typeof total.value !== 'number' || !Number.isFinite(total.value)) {
      return failure('RESPONSE_MAPPING_TYPE_MISMATCH', totalPath);
    }
    data.total = total.value;
  }
  return { ok: true, data };
}

module.exports = { mapResponse };

File: conformance/reference-python/response_mapping.py
import math


def failure(code, path):
    return {"ok": False, "code": code, "path": path}


def read_path(response, path):
    current = response
    for segment in path.split("."):
        if not isinstance(current, dict) or segment not in current:
            return False, None
        current = current[segment]
    return True, current


def resolve_mapped_value(response, path):
    found, value = read_path(response, path)
    if not found:
        return failure("RESPONSE_MAPPING_PATH_MISSING", path)
    return {"found": True, "value": value}


def map_form_record(input_value):
    mapping = input_value.get("responseMapping")
    if mapping is None or not isinstance(mapping, dict) or not mapping:
        return failure("INVALID_RESPONSE_MAPPING", "responseMapping")
    values = {}
    for field, path_expr in mapping.items():
        if not isinstance(field, str) or not field or not isinstance(path_expr, str) or not path_expr:
            return failure("INVALID_RESPONSE_MAPPING", f"responseMapping.{field}")
        found, value = read_path(input_value["response"], path_expr)
        # ADR-0021 D5a / V273: missing path does not abort whole fill.
        # Conformance-observable value is JSON null (equivalent empty initial).
        values[field] = value if found else None
    return {"ok": True, "values": values}


def map_response(input_value):
    if input_value.get("component") == "formRecord":
        return map_form_record(input_value)
    mapping_is_present = "localMapping" in input_value or "datasourceMapping" in input_value
    mapping = (
        input_value["localMapping"]
        if "localMapping" in input_value
        else input_value.get("datasourceMapping")
    )
    if mapping_is_present and (mapping is None or not isinstance(mapping, dict) or not mapping):
        return failure("INVALID_RESPONSE_MAPPING", "localMapping")
    if input_value["component"] in ("statCard", "text") and mapping is not None:
        return failure("RESPONSE_MAPPING_NOT_SUPPORTED", "localMapping")
    if input_value["component"] == "chart" and mapping is None:
        if isinstance(input_value["response"], list):
            return {"ok": True, "data": {"list": input_value["response"]}}
        return failure("RESPONSE_MAPPING_TYPE_MISMATCH", "$")

    list_path = (mapping or {}).get("list", "list")
    list_value = resolve_mapped_value(input_value["response"], list_path)
    if not list_value.get("found"):
        return list_value
    if not isinstance(list_value["value"], list):
        return failure("RESPONSE_MAPPING_TYPE_MISMATCH", list_path)

    data = {"list": list_value["value"]}
    if input_value["component"] == "table" and input_value.get("paginationMode") == "server":
        total_path = (mapping or {}).get("total", "total")
        total_value = resolve_mapped_value(input_value["response"], total_path)
        if not total_value.get("found"):
            return total_value
        total = total_value["value"]
        if isinstance(total, bool) or not isinstance(total, (int, float)) or not math.isfinite(total):
            return failure("RESPONSE_MAPPING_TYPE_MISMATCH", total_path)
        data["total"] = total
    return {"ok": True, "data": data}